### PR TITLE
add output projector option to diffusion models

### PR DESCRIPTION
**Pull request by arrjon**
The output projection layer was incompatible with the transformer backbone in the diffusion type models when employed together with masking of targets. I added an optional `output_projector` parameter to the constructors of these models, allowing users to specify a custom output projection layer. If not provided, a default dense layer is used as before, except when the subnet is a `"diffusion_transformer"`, in which case an identity layer is used.

### CODE REVIEW

**Comment by stefanradev93** [comment]
Since we are now more aware of breaking changes and this is breaking, wouldn't it be better to check in `build()` if the subnet is a diffusion transformer and select identity there? So far, we don't have other incompatible layers.

**Comment by arrjon** [comment]
I removed the `output_projector` and instead check it in the built function now. Trained transformer backbones will still break, but they were only introduced with the last version. Is this acceptable? The way I implemented it now, a user can define a flag in their own subnets to avoid having a dense layer if this is undesired.

**Comment by stefanradev93** [comment]
What if we checked if the subnet output shape is already the same as xz_shape and only if it's not, we add an output projector? We still check for diffusion transformer incompatibility. Should cover all existing cases and most future cases.

**Comment by arrjon** [comment]
Agreed, we can do this. Will still be a breaking change for the transformer backbone (and also on inference tasks where the output of the MLP has the same width as the parameter dimension, unlikely but can happen).

**Comment by arrjon** [comment]
Okay, now it is non-breaking, but I needed to add a `from_config` method.

**Comment by stefanradev93** [comment]
Do we even need the `self._project_output` flag? Why not simply choose between a projector or an identity?

**Comment by arrjon** [comment]
`self._project_output` is only there for backward compatibility and makes it a non-breaking change.

**Comment by stefanradev93** [comment]
But the attribute is not there in `x.13`. A network with an output projector from `x.13/12...` will still load properly. What's the use case for the extra check? Or do you wanna have the extra kwarg as a hidden flag for a special use case?

**Comment by arrjon** [comment]
Older networks that do not have the attribute stored will default to True:
``` return super().from_config({"project_output": True} | config ```

Newer networks will either have the projector or not depending on whether the subnet has the correct dimensions or not. This is the only purpose for the flag. If we removed it and only check for the dimensions, this would be a breaking change, as old networks would just drop their projection layer.

**Comment by arrjon** [comment]
I reverted to mostly non-breaking to keep the code more readable. Diffusion transformer backbone might fail silently and if the MLP width is exactly the parameter dimension.

File: bayesflow/networks/inference/diffusion/diffusion_model.py [hunk 1]
```diff
@@ -268,14 +268,20 @@ def build(self, xz_shape: Shape, conditions_shape: Shape = None):
...
-        units = 1 if self._prediction_type == "potential" else xz_shape[-1]
-        self.output_projector = keras.layers.Dense(units=units, bias_initializer="zeros")
-
         # construct input shape for subnet and subnet projector
         time_shape = (xz_shape[0], 1)
         self.subnet.build((xz_shape, time_shape, conditions_shape))
         out_shape = self.subnet.compute_output_shape((xz_shape, time_shape, conditions_shape))
 
+        units = 1 if self._prediction_type == "potential" else xz_shape[-1]
+        if out_shape[-1] != units:
```
Comment (stefanradev93) [inline bayesflow/networks/inference/diffusion/diffusion_model.py:277]: Isn't this true for potential and creates a projector anyways?

Comment (arrjon) [reply bayesflow/networks/inference/diffusion/diffusion_model.py:277]: Most likely yes, but I think there is no harm in doing it like this. Maybe the subnet already has singluar output for a specific reason...
